packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,401 @@
+    dependent_on_essential: &mut Vec<HashMap<EssentialDataOrigin, Vec<StateVarPointer>>>,
+    essential_data: &mut Vec<HashMap<EssentialDataOrigin, EssentialStateVar>>,
+    freshen_stack: &mut Vec<StateVarPointer>,
+    should_initialize_essential_data: bool,
+) -> () {
+    // This function currently implements recursion through an iterative method,
+    // using a stack on the heap.
+    // This approach was chosen because the function recursion implementation would overflow
+    // the small WASM stack once it recursed a few thousands of times.
+    // An alternative approach would be to increase the size of the WASM stack.
+    // This approach was chosen because it is relatively easy to produce documents
+    // with thousands of levels in the dependency graph, and it wasn't clear what
+    // size WASM stack would be appropriate.
+
+    let original_freshness = components[original_state_var_ptr.component_idx]
+        .borrow()
+        .get_state_variables()[original_state_var_ptr.state_var_idx]
+        .get_freshness();
+
+    // If the current state variable is fresh, there's nothing to do.
+    // If it is unresolved, resolve it
+    match original_freshness {
+        Freshness::Fresh => return (),
+        Freshness::Unresolved => {
+            resolve_state_var(
+                original_state_var_ptr,
+                components,
+                dependencies,
+                dependent_on_state_var,
+                dependent_on_essential,
+                essential_data,
+                should_initialize_essential_data,
+            );
+        }
+        Freshness::Stale | Freshness::Resolved => (),
+    };
+
+    // At this point, the state variable is stale or resolved.
+
+    freshen_stack.push(original_state_var_ptr);
+
+    while let Some(state_var_ptr) = freshen_stack.pop() {
+        let current_freshness = components[state_var_ptr.component_idx]
+            .borrow()
+            .get_state_variables()[state_var_ptr.state_var_idx]
+            .get_freshness();
+
+        // If we have cycles in the graph, it's possible that this
+        // state variable was reached from another path and already freshened.
+        // Skip it in that case.
+        match current_freshness {
+            Freshness::Fresh => continue,
+            Freshness::Unresolved => {
+                panic!("Should not have an unresolved state variable here!")
+            }
+            Freshness::Stale | Freshness::Resolved => (),
+        };
+
+        let dependencies_for_state_var =
+            &dependencies[state_var_ptr.component_idx][state_var_ptr.state_var_idx];
+
+        let mut found_stale_or_resolved_dependency = false;
+
+        for deps in dependencies_for_state_var.iter() {
+            for dep in deps.iter() {
+                match &dep.source {
+                    DependencySource::StateVar {
+                        component_idx,
+                        state_var_idx,
```

Comment:
Since you're only using these to create a new pointer, you could implement `into StateVarPointer` and not de-structure this.

Reply:
I added this trait. The code is now shorter, but seems less explicit. I added some comments to explain what is going on. It seemed unnecessary to have a match since try_from did everything.